v3_async_wip/tests/simple_e2e.py

```diff
@@ -0,0 +1,213 @@
+Will be removed once full E2E support implemented"""
+
+import asyncio
+import logging
+import os
+import pytest
+from dev_utils import iptables
+from v3_async_wip import transport_helper, mqtt_client
+
+logger = logging.getLogger(__name__)
+
+PORT = 8883
+TRANSPORT = "tcp"  # websockets
+IPTABLES_TRANSPORT = "mqtt"  # mqttws
+
+CONNECTION_STRING = os.getenv("IOTHUB_DEVICE_CONNECTION_STRING")
+HOSTNAME = transport_helper.get_hostname(CONNECTION_STRING)
+
+
+class Dropper(object):
+    def __init__(self, transport):
+        self.transport = transport
+
+    def disconnect_outgoing(self, disconnect_type):
+        iptables.disconnect_output_port(disconnect_type, self.transport, HOSTNAME)
+
+    def drop_outgoing(self):
+        iptables.disconnect_output_port("DROP", self.transport, HOSTNAME)
+
+    def reject_outgoing(self):
+        iptables.disconnect_output_port("REJECT", self.transport, HOSTNAME)
+
+    def restore_all(self):
+        iptables.reconnect_all(self.transport, HOSTNAME)
+
+
+@pytest.fixture(scope="function")
+def dropper():
+    dropper = Dropper(IPTABLES_TRANSPORT)
+    yield dropper
+    logger.info("restoring all")
+    dropper.restore_all()
+
+
+@pytest.fixture
+async def client():
+    client_id = transport_helper.get_client_id(CONNECTION_STRING)
+    username = transport_helper.get_username(CONNECTION_STRING)
+    password = transport_helper.get_password(CONNECTION_STRING)
+    ssl_context = transport_helper.create_ssl_context()
+
+    client = mqtt_client.MQTTClient(
+        client_id=client_id,
+        hostname=HOSTNAME,
+        port=PORT,
+        transport=TRANSPORT,
+        keep_alive=5,
+        auto_reconnect=False,
+        ssl_context=ssl_context,
+    )
+    client.set_credentials(username, password)
+    yield client
+    await client.disconnect()
+
+
+def assert_connected_state(client):
+    assert client.is_connected()
+    assert client._desire_connection
+    assert not client._network_loop.done()
+
+
+def assert_disconnected_state(client):
+    assert not client.is_connected()
+    assert not client._desire_connection
+    assert client._network_loop is None
+
+
+def assert_dropped_conn_state(client):
+    assert not client.is_connected()
+    assert client._desire_connection
+    assert client._network_loop.done()
+
+
+@pytest.mark.it("Connect and disconnect")
+async def test_connect_disconnect_twice(client):
+    async def conn_disconn():
+        # Connect
+        await client.connect()
+        assert_connected_state(client)
+        # Wait
+        await asyncio.sleep(1)
+        # Disconnect
+        await client.disconnect()
+        assert_disconnected_state(client)
+
+    # Do it twice to make sure it's repeatable
+    await conn_disconn()
+    await asyncio.sleep(1)
+    await conn_disconn()
+
+
+@pytest.mark.it("Queued connects and disconnects")
+async def test_queued_connects_and_disconnects(client):
+    # TODO: can this be made to show they all happened in order?
```

Comment:
>can this be made to show they all happened in order? [](http://example.com/codeflow?start=12&length=52)

You could assert that they run _serially_, but anything more specific would mean relying on undefined behavior.